### PR TITLE
feat: `gp validate` in Gitpod CLI

### LINKED ISSUE #9025: [gp-cli] Validate .gitpod.yml
## Is your feature request related to a problem? Please describe
Currently, using VSCode, yaml-schema validation helps reducing typos or indentation issues in `.gitpod.yml`. It would be useful to offer the same functionality from the gp CLI.

## Describe the behaviour you'd like
`gp validate` or `gp validate config`

## Describe alternatives you've considered
It'd be interesting if we could also validate `.gitpod.Dockerfile`

## Additional context
N/A

### PR DESCRIPTION
**Pull request by Siddhant-K-code**
Signed-off-by: Siddhant Khare <siddhant@gitpod.io>

## Description

:tada: Validate your `.gitpod.yml` using `gp validate`

|No `.gitpod.yml` file|Valid File|Invalid File _(this can be improved)_|
|:---:|:---:|:---:|
|image|image |image|

### TODO:

- Better error handling of error messages

## Related Issue(s)

Fixes #9025

## How to test

- Open Preview
- Open workspace with any repo.
-  Add a `.gitpod.yml` file:

```yml
# List the start up tasks. Learn more https://www.gitpod.io/docs/config-start-tasks/
tasks:
  - init: echo 'init script' # runs during prebuild
    command: echo 'start script'

# List the ports to expose. Learn more https://www.gitpod.io/docs/config-ports/
ports:
  - port: 3000
    onOpen: open-preview
```
- Open Terminal & run `gp validate`

## Release Notes

```release-note
Validate your .gitpod.yml from Gitpod CLI using `gp validate`
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by akosyakov** [comment]
Do we really need it independent? Or it could be part of the inner loop for instance `gp rebuild`? 

I'm not sure we should add it.

File: components/gitpod-cli/cmd/validate.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+var validateCommand = &cobra.Command{
+	Use:   "validate",
+	Short: "Validate your .gitpod.yml file is valid or not",
+	Run: func(_ *cobra.Command, _ []string) {
+
+		// Load .gitpod.yml schema
+		schemaLoader := gojsonschema.NewReferenceLoader("https://gitpod.io/schemas/gitpod-schema.json")
+
+		// Load .gitpod.yml file
+		filePathOfGitpodConfig := os.Getenv("GITPOD_REPO_ROOT") + "/.gitpod.yml"
```
Comment (akosyakov) [inline components/gitpod-cli/cmd/validate.go:27]: only supervisor can make use of env vars, all other clients should use supervisor, they are implementation details of supervisor

**Comment by Siddhant-K-code** [comment]
> it could be part of the inner loop for instance gp rebuild?

Yeah, we can add this validation step before the rebuild process starts.

**Comment by andreafalzetti** [comment]
> Do we really need it independent? Or it could be part of the inner loop for instance `gp rebuild`? I'm not sure we should add it.

Thanks @Siddhant-K-code for adding this 🙌 

@akosyakov How do you imagine validation to work in rebuild? I imagine the scenario in which I only want to validate my YAML and I don't want to trigger any image build, in this case I would need to pass an additional flag `gp rebuild --validate` which makes the command's DevX drop.

Another idea, could be wrapping `init` and `validate` together, because they both operate on `.gitpod.yml`. 

Overall, I am not against `gp validate` for now, as it follows the current approach which we can revisit when we work on #11371

**Comment by filiptronicek** [comment]
> Another idea, could be wrapping init and validate together, because they both operate on .gitpod.yml.

`gp config init` & `gp config validate`? 👀 

cc @andreafalzetti @Siddhant-K-code

**Comment by akosyakov** [comment]
I mean gp is kind of public API, we are arguing about one command, but one another side accept another command without any design. I personally don't know whether there is a real need for this command and where it should sit properly. I'm not sure we should ad hoc add more commands if we not aligned on design, or we should have something like experimental mode which indicates that this command can go away. The same problem with `gp rebuild` design.

For `gp rebuild` I don't think it needs any arguments, i would think it should fail if parsing cannot happen.

**Comment by Siddhant-K-code** [comment]
Thanks for taking out some time to review this, @akosyakov & @andreafalzetti 🙏🏽 ❤️ 

I think it's good we discuss it first that what our plans and end goals regarding `gp rebuild` & Gitpod CLI Design (#11371), then we can work on this in the future 🔜 

I am closing the PR for now.

**Pull request closed without merging**